server/src/main/java/org/opensearch/common/blobstore/AsyncMultiStreamEncryptedBlobContainer.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.blobstore;
+
+import org.opensearch.common.StreamContext;
+import org.opensearch.common.blobstore.stream.write.WriteContext;
+import org.opensearch.common.io.InputStreamContainer;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.crypto.CryptoManager;
+
+import java.io.IOException;
+
+/**
+ * EncryptedBlobContainer is an encrypted BlobContainer that is backed by a
+ * {@link AsyncMultiStreamBlobContainer}
+ *
+ * @opensearch.internal
+ */
+public class AsyncMultiStreamEncryptedBlobContainer extends EncryptedBlobContainer implements AsyncMultiStreamBlobContainer {
+
+    private final AsyncMultiStreamBlobContainer blobContainer;
+    private final CryptoManager cryptoManager;
+
+    public AsyncMultiStreamEncryptedBlobContainer(AsyncMultiStreamBlobContainer blobContainer, CryptoManager cryptoManager) {
+        super(blobContainer, cryptoManager);
+        this.blobContainer = blobContainer;
+        this.cryptoManager = cryptoManager;
+    }
+
+    @Override
+    public void asyncBlobUpload(WriteContext writeContext, ActionListener<Void> completionListener) throws IOException {
+        EncryptedWriteContext encryptedWriteContext = new EncryptedWriteContext(writeContext, cryptoManager);
+        blobContainer.asyncBlobUpload(encryptedWriteContext, completionListener);
+    }
+
+    @Override
+    public boolean remoteIntegrityCheckSupported() {
+        return false;
+    }
+
+    static class EncryptedWriteContext extends WriteContext {
+
+        private final Object encryptionMetadata;
+        private final CryptoManager cryptoManager;
+        private final long fileSize;
+
+        /**
+         * Construct a new encrypted WriteContext object
+         */
+        public EncryptedWriteContext(WriteContext writeContext, CryptoManager cryptoManager) {
+            super(writeContext);
+            this.cryptoManager = cryptoManager;
+            this.encryptionMetadata = cryptoManager.initEncryptionMetadata();
+            this.fileSize = cryptoManager.estimateEncryptedLengthOfEntireContent(encryptionMetadata, writeContext.getFileSize());
+        }
+
+        public StreamContext getStreamProvider(long partSize) {
```

Comment:
Can we utilize or configure the `preferredLength` as a part of the repository properties?
This will ensure the same part size is used at the time of upload, download as well as any encryption/decryption logic since the part size has to match across for partial file crypto logic.

Reply:
partSize is not a constant. It is based on the size of the file.